### ISSUE TITLE
Decrypt problem with CMS Enveloped data (padding issue?)

### ISSUE THREAD

**Issue by matejsp** [bcgit/bc-java#771]
We are using CMS with BC 1.66 using X509 certificate with RSA private key and are trying to encrypt the payload.
This example uses payload of 8704 * 'A' ... If you try with a smaller number it works.

However when using larger payload you get:

Stacktrace:
```
Exception in thread "main" org.bouncycastle.crypto.io.InvalidCipherTextIOException: Error finalising cipher
	at org.bouncycastle.jcajce.io.CipherInputStream.finaliseCipher(Unknown Source)
	at org.bouncycastle.jcajce.io.CipherInputStream.close(Unknown Source)
	at java.base/java.io.BufferedInputStream.close(BufferedInputStream.java:489)
	at java.base/java.io.FilterInputStream.close(FilterInputStream.java:180)
	at net.iconomi.halcomb2b.util.CMSCryptographyUtils.decrypt(CMSCryptographyUtils.java:157)
	at net.iconomi.halcomb2b.util.TestBCError.main(TestBCError.java:44)
Caused by: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
	at java.base/com.sun.crypto.provider.CipherCore.unpad(CipherCore.java:975)
	at java.base/com.sun.crypto.provider.CipherCore.fillOutputBuffer(CipherCore.java:1056)
	at java.base/com.sun.crypto.provider.CipherCore.doFinal(CipherCore.java:853)
	at java.base/com.sun.crypto.provider.AESCipher.engineDoFinal(AESCipher.java:446)
	at java.base/javax.crypto.Cipher.doFinal(Cipher.java:2085)
	... 6 more
```

Code:
```
import org.bouncycastle.cms.CMSAlgorithm;
import org.bouncycastle.cms.CMSEnvelopedDataParser;
import org.bouncycastle.cms.CMSEnvelopedDataStreamGenerator;
import org.bouncycastle.cms.CMSException;
import org.bouncycastle.cms.Recipient;
import org.bouncycastle.cms.RecipientInformation;
import org.bouncycastle.cms.jcajce.JceCMSContentEncryptorBuilder;
import org.bouncycastle.cms.jcajce.JceKeyTransEnvelopedRecipient;
import org.bouncycastle.cms.jcajce.JceKeyTransRecipientInfoGenerator;
import org.bouncycastle.operator.OutputEncryptor;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.PrivateKey;
import java.security.UnrecoverableKeyException;
import java.security.cert.CertificateEncodingException;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.Collection;
import java.util.Iterator;

public class TestBCError {

    public static void main(String[] args) throws KeyStoreException, CertificateException, NoSuchAlgorithmException, IOException, UnrecoverableKeyException, CMSException {

        char[] password = "service1password".toCharArray();

        InputStream is = TestBCError.class.getResourceAsStream("/service1.p12");

        KeyStore ks = KeyStore.getInstance("PKCS12");
        ks.load(is, password);

        X509Certificate service1cert = (X509Certificate) ks.getCertificate("service1");
        PrivateKey service1privateKey = (PrivateKey)ks.getKey("service1", password);

        byte[] data = encrypt(service1cert, repeat('A', 8704).getBytes());
        data = decrypt(service1privateKey, data);
    }

    private static RecipientInformation getSingleRecipient(CMSEnvelopedDataParser parser) {
        Collection recInfos = parser.getRecipientInfos().getRecipients();
        Iterator recipientIterator = recInfos.iterator();
        if (!recipientIterator.hasNext()) {
            throw new RuntimeException("Could not find recipient");
        }
        return (RecipientInformation) recipientIterator.next();
    }

    public static byte[] encrypt(X509Certificate certificate, byte[] unencrypteddata) throws CertificateEncodingException, CMSException, IOException {
        CMSEnvelopedDataStreamGenerator gen = new CMSEnvelopedDataStreamGenerator();
        gen.addRecipientInfoGenerator(new JceKeyTransRecipientInfoGenerator(certificate));
        OutputEncryptor encryptor = new JceCMSContentEncryptorBuilder(CMSAlgorithm.AES256_CBC).build();

        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
             OutputStream encryptingStream = gen.open(baos, encryptor)) {
            encryptingStream.write(unencrypteddata);
            encryptingStream.close();
            return baos.toByteArray();
        }
    }

    public static byte[] decrypt(PrivateKey privateKey, byte[] encryptedData) throws IOException, CMSException {
        CMSEnvelopedDataParser parser = new CMSEnvelopedDataParser(encryptedData);
        RecipientInformation recInfo = getSingleRecipient(parser);
        Recipient recipient = new JceKeyTransEnvelopedRecipient(privateKey);

        try (InputStream decryptedStream = recInfo.getContentStream(recipient).getContentStream()) {
            return decryptedStream.readAllBytes();
        }
    }

    public static String repeat(final char ch, final int repeat) {
        if (repeat <= 0) {
            return "";
        }
        final char[] buf = new char[repeat];
        for (int i = repeat - 1; i >= 0; i--) {
            buf[i] = ch;
        }
        return new String(buf);
    }

}
```

**Comment by matejsp** [comment]
I have found a workaround. Fix is to replace:
```decryptedStream.readAllBytes();```
with
```IOUtils.toByteArray(decryptedStream)```

Both methods (internal java and commons-io) use the same DEFAULT_BUFFER_SIZE = 8192.
It seems that readAllBytes is corrupting somehow the internal state of bouncy castle CipherInputStream.

**Comment by martijnbrinkers** [comment]
This looks like a bug in org.bouncycastle.cms.CMSTypedStream$FullReaderStream#read

```
        public int read(byte[] buf, int off, int len) throws IOException
        {
            int totalRead = Streams.readFully(super.in, buf, off, len);
            return totalRead > 0 ? totalRead : -1;
        }
    }
```

if len == 0, totalRead will be 0 and therefore -1 is returned indicating that the EOF was reached which then results in readAllBytes() from reading more data. The readAllBytes method contains some "weird" code that does a read with 0 length. Although weird, in principle that should be okay.

I think CMSTypedStream$FullReaderStream#read should return 0 if len is 0.

**Comment by matejsp** [comment]
Nice catch. You are correct. This is issue in FullReaderStream.read ...

Second read is called with 0 length.

https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read(byte[],%20int,%20int)

> If len is zero, then no bytes are read and 0 is returned; otherwise, there is an attempt to read at least one byte. If no byte is available because the stream is at end of file, the value -1 is returned; otherwise, at least one byte is read and stored into b.

**Comment by borissnd** [comment]
I ran into this issue as well. fix seems simple enough. would it be possible to get this prioritized please?

**Comment by dghgit** [comment]
Thanks for the report. Closing as fixed in 1.68 and somehow never got around to updating the jira.

@borissnd if you are still seeing an issue and it's with a release later than 1.68, please raise a new issue.
